### PR TITLE
Make locator reusable

### PR DESCRIPTION
**Pull request by Stadly**
Extract functionality that is common for all locator classes.

### CODE REVIEW

**Comment by shadowhand** [comment]
Why?

File: src/DiscoveryLocator.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+{
+    /** @var array  */
+    protected static $candidates = [];
+
+    /**
+     * @throws RuntimeException If no implementation is available
+     */
+    public static function locate(string $interface): string
+    {
+        foreach (static::$candidates[$interface] ?? [] as $candidate) {
```
Comment (shadowhand) [inline src/DiscoveryLocator.php:19]: No need for `??` here, the type is declared as an array and has a default value.

**Review by shadowhand — CHANGES_REQUESTED** [review]

File: src/DiscoveryLocator.php (hunk 1 above)
Comment (Stadly) [reply src/DiscoveryLocator.php:19]: If you remove `?? []`, the test `testCannotLocateInvalidFactory` will fail. It is a public method, so we cannot assume that `$interface` actually is a key in the array.

Comment (shadowhand) [reply src/DiscoveryLocator.php:19]: Ah, right. Still working on my first cup of tea. 😄

**Review by shadowhand — APPROVED** [review]

**Comment by shadowhand** [comment]
Thanks!

**Pull request merged**